### ISSUE TITLE
Feature request: better C++ support

### ISSUE THREAD

**Issue by darktjm** [panda3d/panda3d#1518]
I apologize to fans out there, but I despise Python with the passion of a thousand fiery suns.  I originally avoided panda3d due to its official stance (stated in multiple places) that it's an engine for Python game development.  However, after finding that the C++ interface is actually usable (even moreso than some other pure C/C++ engines), I have decided to try using panda3d for a project.

There are a number of obstacles, some of which may be my own imagination due to the fact that I'm new to panda3d:

 - Almost all documentation, even the official (on-line only, which is another thing I don't care for, but that's a separate issue) "C++" documentation, which is focused entirely on the Python API.  About the only thing I have done in this regard is try to coax a doxygen document out of it.  I don't care for doxygen's output formats, and the fact that a lot of documentation is in the implementation files (.I and .cxx) rather than the API files (.h) makes it very difficult, but it's better than nothing.  Any links to documentation I have found have been dead or Python-only.  The only scripts to generate API documentation in 1.10.13 are python2 scripts I can't even run and don't want to port.
 - There are no C++ samples.  I have at least partially worked on this (https://github.com/darktjm/c-p3d-samples), but I am new to panda3d and probably didn't do it the "right" way.  In fact, during development, I made a few major changes based on new discoveries.
 - There are some useful features only available in the Python API.  There are a few separate issues already in that regard (e.g. #1508), so I suppose that this should be a separate list of specific features.  Probably the only ones I've encountered so far (but I may be wrong, since I'm new to panda3d) are the BufferViewer and Filters classes.
 - There is almost no C++ build support in 1.10.13.  I know the CMake port is ongoing, and this will fix it a little.  Maybe also a pkg-config support (#788) and, if not already present, a FindPackage module for panda3d.
 
This is a very general request, and should probably be split into separate issues, I guess, but I'm not even sure if the current project maintainers feel this is a good idea.  Feel free to close it as "too vague".

**Comment by ArsThaumaturgis** [comment]
Regarding the first point at least, in the manual there should be a link near the top-left, just below the version-dropdown, to switch to the C++ version of the manual and API.

**Comment by serkkz** [comment]
In my opinion, point 3 is not a problem, since the application template is not the fundamental capabilities of the engine. My thought is that the template should depend on the goals of the application, for example, you want to create an fps or rts game. Accordingly, the functionality of one of the templates may be unclaimed.

What should the user do if he does not need aspect2d child nodes, he uses the pixel2d node, and moreover, he does not need unnecessary calculations of transformation into nodes, the user categorically objects to any meaningless background CPU consumption.

The fourth point is not true, you can build Panda3D from the source code with the necessary libraries and plugins. It doesn't matter if you use CMake or MakePanda. At the moment, the github build is positioned as an SDK, but no one forbids you to build your application with the necessary options.

**Comment by darktjm** [comment]
> Regarding the first point at least, in the manual there should be a link near the top-left, just below the version-dropdown, to switch to the C++ version of the manual and API.

Are you referring to https://docs.panda3d.org/1.10/cpp/index?  Or is there another manual I should look for?  Because that's what I meant by "The official documentation".  Maybe my whole trouble is that I was looking in the wrong place.

As to that manual:  I looked at a few specific things, found them to be exact or nearly exact duplicates of the Python pages, including the use of Python, so I stopped reading.  Even the API document appeared at first glance to be Python, since you use subsystem names that are only present in the Python API.  If it weren't an on-line-only document, I might have given it more thought, but I have no patience for that sort of crap.

For example, the first thing I ran into is that I wanted to make feature parity on the ball-in-maze sample between the one I found out there and the Python sample.  Here's the page on interval animations, which has a starting paragraph or two about the C++ interface, but then reverts entirely into Python:

https://docs.panda3d.org/1.10/cpp/programming/intervals/index

In fact, it was only after my first attempt, using tasks, that I discovered that CInterval exists and can sort of be used, with a little extra help.  This is because I looked for "Sequence", found the task sequence, and used it, rather than using CMetaInterval.

The GUI documentation covers only the Python DirectGUI, and doesn't even mention PGUI (as far as I can tell; I'm not going to search an on-line only document for hours).  The particle system documentation seems to start with C++ stuff, but for some things (such as loading particle systems, which as far as I can tell constitutes loading and interpreting raw Python files), the documentation silently reverts to Python.  The physics system documentation starts out with a Python-only API call.  And so forth.  Bite my hand often enough, and I won't extend it any more.

Make it possible for me to download a manual (if it's not already possible), preferably in source format, and not only will I read it, I'll correct issues as I find them.  As is, you need to replace the duplication of Python documentation with "This feature is only available in Python" or "This is the pain-in-the-ass method you have to use in C++, because in reality, the C++ is only there to accelerate the Python API, or was added as an afterthought".  Or maybe say it in a less abrasive way, your choice.

I will grant you that there is a complete C++ example buried in the documentation, but as I said, I abandoned that early on after discovering too much Python in it.  I guess I should read that example now that I'm aware of it.  I'm almost done porting the Python samples over to C++ now, anyway, though.  Just 5 more to go (after abandoning a few because porting the Python library is too much trouble), but little to no motivation any more.

> What should the user do if he does not need aspect2d child nodes

Not enable it?  Like other things that are auto-created on first use?

If you want to use intervals, you have to create a task to step them.   Trivial code, but having a function to do so (possibly automatically) is what makes the Python API better.  Code bloat?  In this day an age, that's not an issue.  See the bottom of my c-panda-samples README for a longer list of missing features.

> The fourth point is not true, you can build Panda3D from the source 

So?  I build panda3d from git, anyway (albeit from a tagged release point).  How does that help in building my own applications?  To clarify:  `$(pkg-config panda3d --cflags --libs)` or, in CMake, `find_package(Panda3D REQUIRED)` rather than manually searching for potential SDK install locations (and guessing at what additional libraries may be required, or what set of p3d libraries to link with) is what I meant by build support.  Perhaps you're working on it, but I refuse to use git head without very good reason (e.g. major bug fixes), especially when I can't predict the next release (time or contents).

**Comment by serkkz** [comment]
> So? I build panda3d from git, anyway (albeit from a tagged release point). How does that help in building my own applications? To clarify: `$(pkg-config panda3d --cflags --libs)` or, in CMake, `find_package(Panda3D REQUIRED)` rather than manually searching for potential SDK install locations (and guessing at what additional libraries may be required, or what set of p3d libraries to link with) is what I meant by build support. Perhaps you're working on it, but I refuse to use git head without very good reason (e.g. major bug fixes), especially when I can't predict the next release (time or contents).

If you talk about it in this way, I partially agree. Indeed, the average user does not know the internal components of the engine, and it is not always obvious what to include in the build. However, this is partially stated here:

https://docs.panda3d.org/1.10/python/distribution/thirdparty-licenses

The fact that you refuse to find out anything about the necessary libraries does not mean that it is impossible to build your application using CMake.

The first step is to build the engine with the library components that are required for your project.
The second stage is you build your application as usual using something like this.

```cmake
cmake_minimum_required(VERSION 3.22)

# The name of the project.
project(Game VERSION 1.0)

# Directory where the Panda3D SDK is installed.
set(PandaSDK D:/Panda3D)

link_directories(Game ${PandaSDK}/lib)

# Name of the executive file.
add_executable(Game main.cxx)

# Set the dependency for the Game execution file.
target_link_libraries(Game libp3framework)
target_link_libraries(Game libpanda)
target_link_libraries(Game libpandaexpress)
target_link_libraries(Game libp3dtool)
target_link_libraries(Game libp3dtoolconfig)

# Specify where to look for Panda3D headers.
include_directories(Game ${PandaSDK}/include)
```

**Comment by darktjm** [comment]
Sorry, to clarify, when I say "you" I mean "you" in the plural, general sense, formerly commonly expressed as "you guys" or "you people" for clarity, but some people are offended by that.  I meant "whoever feels like actually addressing this issue" aka "the people responsible".  I am obviously not one of those people, although having a better way to retrieve and edit documentation would make me more likely to fix some of the documentation issues.  My only "contribution" to date is the aforementioned sample ports, https://github.com/darktjm/c-p3d-samples.

Also, to further clarify:  I now admit I was wrong about point #2, which neither of you specifically addressed.  There is an official tutorial sample in the C++ documentation, but I missed it because I stopped reading too soon.  I still think it's worthwhile to port the official sample collection supplied with the panda3d distribution, though, rather than relying on an external site.  Again, my complaint is partially that it's on-line-only.  Is there a place I can download this sample, like the solar-system sample?  Or am I just expected to type it in (yeah, I know, it's a tutorial, so probably I should just type it in).  In any case, I'll finish my sample porting as is first before reading the tutorial and correcting everything I did wrong.  Maybe I'll even read the full C++ documentation and collect some notes about missing/wrong bits and look over the link of documentation issues.  I'm tired of this preliminary work, though, and want to concentrate on the project I originally evaluated panda3d for.  I've already identified a few missing panda3d features common to other engines I will have to implement myself.  That way, if panda3d ends up being more trouble than it's worth, I can hold my nose and go with Godot, which is about the only engine I haven't given any serious thought to yet (I don't see any way of separating out the Godot's Own Python-like Programming Language and Mandatory Launcher), but is popular enough that at least I won't feel like it hasn't been updated in years and might suddenly be abandoned just like many popular engines of the past.

Sorry for rambling, especially with info that isn't relevant to the issue.

**Comment by darktjm** [comment]
> The fact that you refuse to find out anything about the necessary libraries does not mean that it is impossible to build your application using CMake.

Whatever.  I have my own [CMake config](https://github.com/darktjm/c-p3d-samples/blob/main/supt/samples.cmake), which is both more portable and more complete than what you show.  That doesn't change the fact that it needs to be supplied in the SDK.  This isn't about me.  I've been doing software development for over 30 years, so I can get anything working, even if it's not the "right way".  It's about me and other people wanting to use panda3d for C++ development.  Making everybody reinvent the wheel is not the right way to do things.

**Comment by serkkz** [comment]
Here you need to decide who you want to be - developers or a Panda3D user. Usually users require a button to build the application.

**Comment by ArsThaumaturgis** [comment]
> Sorry, to clarify, when I say "you" I mean "you" in the plural, general sense, formerly commonly expressed as "you guys" or "you people" for clarity, but some people are offended by that. I meant "whoever feels like actually addressing this issue" aka "the people responsible".

I see what you're saying, but as written, especially in a reply, it does feel more directed than that, I fear.

> I now admit I was wrong about point https://github.com/panda3d/panda3d/pull/2, which neither of you specifically addressed.

I'm glad that there are such samples!

Speaking for myself, I didn't respond to that simply because I don't really use the C++ side much; I'm primarily a Python developer these days.

(I used to use C++ many years ago, but I found that I actually preferred Python and haven't really looked back. Each to their own, and all that!)

>  Again, my complaint is partially that it's on-line-only. Is there a place I can download this sample, like the solar-system sample?

Hmm... I haven't tried this myself, but it _looks_ like the manual should be downloadable from its GitHub page.

That is, here:
https://github.com/panda3d/panda3d-docs

I have no idea whether that will produce a usable manual, but it seems worth a try!

**Comment by darktjm** [comment]
> That is, here: https://github.com/panda3d/panda3d-docs

Thanks.  I'll look at that later.  I don't know why I didn't see that before.  Probably too much stuff to start with at once (I'm sticking with that excuse).  It'll probably take me a while just to build it, since I don't have all of the sphinx extensions installed.  *sigh*

**Comment by WMOkiishi** [comment]
For what it's worth, I do know that one of the longer-term goals for Panda3D is to essentially make the `direct` package (where all the Python code is) obsolete. This means improving the parts written in C++ so that they're just as easily usable as the Python stuff build around them in `direct`. If you want to help out with that, [the Panda3D Discord](https://discord.gg/UyepRMm) is probably a good place to get some direction.

**Comment by darktjm** [comment]
> I'm glad that there are such samples!

Actually, I just noticed that the official tutorial example post-links to your (I mean singular this time) tutorial.  I suppose that once I'm done porting (or rejecting for now, as is the case with at least the fireball sample) the last 5 official samples, if ever, I could take a look at porting that to C++.  I didn't look at the code yet, but I assume it's Python only right now.  Naturally that would also involve "porting" the associated documentation.  Then again, I don't know why I feel that's my duty, since I might end up using something else, anyway (raylib or osg for simplicity or godot for insanity).

**Comment by darktjm** [comment]
> This means improving the parts written in C++ so that they're just as easily usable as the Python stuff build around them in `direct`.

That's good to know.  Long-term goals don't help me a lot right now, though.  My project has to get moving.

> If you want to help out with that

Maybe if I end up choosing panda3d.  However, choosing also involves minimizing excess effort, which helping out would definitely be.  In fact, porting all of those samples so far has been more effort than I wanted to put into it right now.  What I wanted was a 3D library that could do the things I wanted without getting in the way of core development.  Thanks to a recent wide-spread week-long power outage and other issues, this is already taking much longer than expected.

> .. Discord ..

Sorry, I currently don't use Discord as a matter of personal policy.

**Comment by serkkz** [comment]
@darktjm 

I visited your repository, read your story about the lack of support for C++, which I partly agree with. But your idea to equalize C++ and python will not lead anywhere, so that's not the problem, the problem is, as you noted, the popularization of python, respectively, there are few users who want to get involved with C++, at the moment there is no problem with any restriction on the C++.

If you want to be able to use Panda3D for your own needs on the C++ side, you need to study panda more deeply. At the moment, the Direct module is a bicycle.

https://github.com/lettier/3d-game-shaders-for-beginners

Here is an example for your piggy bank with C++ code for Panda3D, note that it is implemented without any python functions. I will even say more that the use of the PandaFramework class can be abandoned without any consequences.

**Comment by ArsThaumaturgis** [comment]
> Thanks. I'll look at that later. I don't know why I didn't see that before.

In all fairness, it's _not_ terribly obvious, I feel.



> I suppose that once I'm done porting (or rejecting for now, as is the case with at least the fireball sample) the last 5 official samples, if ever, I could take a look at porting that to C++.

You're welcome to do so! Indeed, it might be nice to have a C++ version of the tutorial for those who are interested in working with that language! ^_^

> I didn't look at the code yet, but I assume it's Python only right now.

You are quite correct.

> Naturally that would also involve "porting" the associated documentation. Then again, I don't know why I feel that's my duty, since I might end up using something else, anyway (raylib or osg for simplicity or godot for insanity).

And that's fair, I do think.

Indeed, if Panda isn't suited to you, then... that's okay, as far as I see. I daresay that _no_ engine is congenial to _all_ developers, so it's a matter of finding whichever suits you best! If that's Panda, then great! If not, then fair enough!

> Thanks to a recent wide-spread week-long power outage and other issues ...

Oof, I'm sorry to read of it! :(

**Comment by darktjm** [comment]
> > I suppose that once I'm done porting (or rejecting for now, as is the case with at least the fireball sample) the last 5 official samples, if ever, I could take a look at porting that to C++.
> 
> You're welcome to do so! Indeed, it might be nice to have a C++ version of the tutorial for those who are interested in working with that language! ^_^

[Well, I tried.](https://github.com/darktjm/c-p3d-samples/tree/main/Panda3DTutorial.io)  Got stuck and dropped my desire to make it a standalone tutorial duplicating most of your documentation and adjusting it for C++, and instead wrote it as a supplement.  Got stuck again porting #15; GUI is the reason I dropped several gamepad samples as well.  Got bored with docs around #8, but pushed through to #11, so #12-14 still need docs, #15 still needs porting, and #16 ain't gonna happen (unless I can think of a suitable replacement procedure).  And the whole thing needs auditing, or at least spell checking.  I might at least push through to #15 some day, but my project partner doesn't really want panda3d, and given that I provided the source material for most of his complaints, I can hardly convince him otherwise.  This has mostly just been me wasting time because I'm insane and trying to avoid other, more pressing issues.

I also found https://github.com/drivird/drunken-octo-robot, making me wonder why Panda3D hasn't had C++ samples for at least 10 years, saving me the effort and Python exposure.  Or at least a link to a C++ example or tutorial at the end of their own C++ tutorial, instead a link to a Python one.  Oh well.  I think I should just close this issue, as well as my other two issues I opened here.

**Comment by ArsThaumaturgis** [comment]
Well, thank you for the work that you did do on the tutorial! Even if partial, perhaps it'll be of aid to someone else! ^_^

As to the repository that you mention, I don't think that I'm familiar with that one--although that's not necessarily indicative.

I suppose then that I see two plausible reasons that it might not have been taken on, offhand: First, that it simply never became known to the community at large, and second, that since I gather that most who employ Panda do so via Python, there was never much motivation to do much with it. But those are guesses!

Anyway, if this means that you're not taking up Panda, then fair enough, and I hope that you find an engine that suits you well! ^_^

**Comment by rdb** [comment]
I think these are good suggestions, and I appreciate you bringing them up.

There's a doxygen version of the Panda3D C++ API, but it's a little hard to integrate it with the new Sphinx-based manual.
https://www.panda3d.org/reference/cxx/

I'm open to the idea of C++ samples, please open an issue for that. There have been a few efforts on this but I have not seen a PR or issue here with a plan for integrating them into the main repository.

**Comment by darktjm** [comment]
I will not be able to log into github any more as of tomorrow, for failure to comply with their 2FA policy. I can't track or comment on this issue any more except via email, and I'm not sure that has ever worked right for me. It was always too generic a request, anyway.  There are probably better ways to track such things.  I put some effort into fixing the first two points.  However, I don't really like the git style of fork+branch+pull request (thus my doc updates are lost to you, and I won't post a request to add my or anyone else's C++ samples, not that I will be able to any more after tomorrow, anyway), and I don't see the point of doing it all myself with little to no benefit, and potentially in conflict with others' efforts.  I'd close this issue, but I will let someone else do it.

**Comment by maxrdz** [comment]
I'm a bit late to the conversation, but I would like to show my support to increasing support for the C++ interface of Panda, or at least moving more performance-demanding tasks from DIRECT over to the C++ end.

Hopefully #1508 helped a bit towards this goal. lol :P
